asserts/authority_delegation.go

```diff
@@ -0,0 +1,166 @@
+	res := make([]*AssertionConstraints, 0, 1)
+	for _, ac := range ad.assertionConstraints {
+		if ac.Check(a) == nil {
+			res = append(res, ac)
+		}
+	}
+	if len(res) == 0 {
+		return nil
+	}
+	return res
+}
+
+// Implement further consistency checks.
+func (ad *AuthorityDelegation) checkConsistency(db RODatabase, acck *AccountKey) error {
+	if !db.IsTrustedAccount(ad.AuthorityID()) {
+		// XXX if this is relaxed then authority-id must otherwise
+		// match account-id
+		return fmt.Errorf("authority-delegation assertion for %q is not signed by a directly trusted authority: %s", ad.AccountID(), ad.AuthorityID())
+	}
+	_, err := db.Find(AccountType, map[string]string{
+		"account-id": ad.AccountID(),
+	})
+	if IsNotFound(err) {
+		return fmt.Errorf("authority-delegation assertion for %q does not have a matching account assertion", ad.AccountID())
+	}
+	if err != nil {
+		return err
+	}
+	_, err = db.Find(AccountType, map[string]string{
+		"account-id": ad.DelegateID(),
+	})
+	if IsNotFound(err) {
+		return fmt.Errorf("authority-delegation assertion for %q does not have a matching account assertion for delegated %q", ad.AccountID(), ad.DelegateID())
+	}
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// sound
+var _ consistencyChecker = (*AuthorityDelegation)(nil)
+
+// Prerequisites returns references to this authority-delegation's prerequisite assertions.
+func (ad *AuthorityDelegation) Prerequisites() []*Ref {
+	return []*Ref{
+		{Type: AccountType, PrimaryKey: []string{ad.AccountID()}},
+		{Type: AccountType, PrimaryKey: []string{ad.DelegateID()}},
+	}
+}
+
+func assembleAuthorityDelegation(assert assertionBase) (Assertion, error) {
+	// account-id and delegate-id are checked by the general
+	// primary key code
+
+	cs, ok := assert.headers["assertions"]
+	if !ok {
+		return nil, fmt.Errorf("assertions constraints are mandatory")
+	}
+	csmaps, ok := cs.([]interface{})
+	if !ok {
+		return nil, fmt.Errorf("assertions constraints must be a list of maps")
+	}
+	if len(csmaps) == 0 {
+		// there is no syntax producing this scenario but be robust
+		return nil, fmt.Errorf("assertions constraints cannot be empty")
+	}
+	acs := make([]*AssertionConstraints, 0, len(csmaps))
+	for _, csmap := range csmaps {
+		m, ok := csmap.(map[string]interface{})
+		if !ok {
+			return nil, fmt.Errorf("assertions constraints must be a list of maps")
+		}
+		typeName, err := checkNotEmptyStringWhat(m, "type", "constraint")
+		if err != nil {
+			return nil, err
+		}
+		t := Type(typeName)
+		if t == nil {
+			return nil, fmt.Errorf("%q is not a valid assertion type", typeName)
+		}
+		hm, err := checkMapWhat(m, "headers", "constraint")
+		if err != nil {
+			return nil, err
+		}
+		cc := compileContext{
+			opts: &compileAttrMatcherOptions{},
+		}
+		matcher, err := compileAttrMatcher(cc, hm)
+		if err != nil {
+			return nil, fmt.Errorf("cannot compile \"headers\" constraint: %v", err)
```

Comment:
Nitpick,
```suggestion
			return nil, fmt.Errorf(`cannot compile "headers" constraint: %v`, err)
```

Reply:
I just removed the "" around headers, we are not super consistent about this anyway